DM-45429: Initial implementation of a general query result

andy-slac opened the pull request:
Registry method `queryDatasetAssociations` is reimplemented (for both
direct and remote butler) to use new query system and new general query
result class.


## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

andy-slac committed:
Fix query joiner construction to avoid cartesian joins.

The code was joining calib table twice, with and without alias, which
caused a cartesian join with that table.

andy-slac committed:
Improve handling of Timespan in queries.

Postgres sometimes needs help in guessing column type when doing UNION
of several queries.

andy-slac committed:
Initial implementation of a general query result (DM-45429)

Registry method `queryDatasetAssociations` is reimplemented (for both
direct and remote butler) to use new query system and new general query
result class.

andy-slac commented:
@TallJimbo, I have an initial version of a general query result calss and I reimplemented `queryDatasetAssociations` to use that to make sure it can work. Could you check that what I've done so far matches your model for it?

TallJimbo commented on python/lsst/daf/butler/queries/_general_query_results.py:
```diff
@@ -0,0 +1,125 @@
+
+    def iter_refs(self, dataset_type: DatasetType) -> Iterator[tuple[DatasetRef, dict[ResultColumn, Any]]]:
+        """Iterate over result rows and return DatasetRef constructed from each
+        row and an original row.
```
There's one place in QG generation where we'd want a method sort of like this that yields tuples of `DatasetRef` and `DataCoordinate` (it currently calls `DataCoordinateQueryResults.findRelatedDatasets`).

I had originally planned to do that via a special dict-like row object that had single-row `get_ref`, `get_data_id` etc. methods, but I think I like your iteration adapter better as it allows for doing work common to all rows up front.  But I'd like to extend this to be able to support at least `DatasetRef` + `DataCoordinate` + [extra stuff] somehow, and maybe even support `DatassetRef` objects for multiple dataset.  Maybe something like this?
```
def iter_tuples(self, *dataset_type: DatasetType) -> Iterator[tuple[DataCoordinate, dict[str, DatasetRef], dict[ResultColumn, Any]]
```
where it always returns the full data ID, a dict of `DatasetRef` keyed by dataset type name, and then whatever columns aren't included in one of those?  To be honest I'm not thrilled with this, either, but maybe it'll inspire you to find something better?

andy-slac replied:
The return annotation is hard to read. Would it make sense to introduce a helper NamedTuple like this:
```
class GeneralResultTuple(typing.NamedTuple):

    data_id: DataCoordinate
    refs: dict[str, DtasetRef]  # or maybe just list[DatasetRef]?
    raw_row: dict[str, Any]

class GeneralQueryResults(QueryResultsBase):

    def iter_tuples(self, *dataset_type: DatasetType) -> Iterator[GeneralResultTuple]:
        ...
```

TallJimbo replied:
:+1: I do like `NamedTuple` for this; you can still unpack it if you want to, but it's also very clear what's in it and you can hang documentation on it.  And I like `list` rather than `dict` for the `DatasetRefs`, too, since we can make that ordering consistent with the argument rather than arbitrary.

TallJimbo commented on python/lsst/daf/butler/queries/_general_query_results.py:
```diff
@@ -0,0 +1,125 @@
+        Yields
+        ------
+        row_dict : `dict` [`ResultColumn`, `Any`]
+            Result row as dictionary, the keys are `ResultColumn` instances.
```
Another possibility for the key type here is a `<dimension-or-dataset-type-name>.<column>` string; that might be better if we make this type more publicly accessible.  In either case it should probably match however we let the user specify the desired result columns in any public interface that returns one of these.

andy-slac replied:
I switched to `str` as a key, it is indeed more natural as API uses strings as parameters. I also replaced `dataset_associations` with `general` which takes the same parameters as `GeneralResultSpec`.

TallJimbo reviewed (commented):
As I mentioned in the stand-up, I had imagined adding a to-be-public `Query.general` method rather than `Query.dataset_associations`, as sort of a power-user escape valve.  I'm not opposed to also having `Query.dataset_associations`.

andy-slac committed:
Switch hybrid registry to use remote registry for queryDatasetAssociations.

This enables testing of the remote registry implementation, which uncovered
a problem with deserialization of general result pages. Had to add additional
code to serialize/deserialize that thing correctly.

andy-slac committed:
Remove `Query.dataset_associations` and replace it with `general` method.

General result iterator now returns dictionaries indexed by strings.

TallJimbo commented on python/lsst/daf/butler/queries/_query.py:
```diff
@@ -292,6 +298,45 @@ def dimension_records(self, element: str) -> DimensionRecordQueryResults:
+        self,
+        dimensions: DimensionGroup,
+        dimension_fields: Mapping[str, set[str]] = {},
+        dataset_fields: Mapping[str, set[DatasetFieldName]] = {},
```
Since we don't need to mutate them, better to make the value types `collections.abc.Set` rather than `set`.

What do you think about making all of these keyword-only arguments, and adding a `*args: str`  variadic argument that we'd filter though `interpret_identifier` (in `queries/_identifiers.py`) and use to update the existing categorized values?

It might also be nice to have a simple way to say "all of the fields necessary for `DatasetRefs` of this dataset type", maybe `datasets={name: ...}` or something?

andy-slac committed:
General query result returns tuples from `iter_tuples` method.

Special NamedTuple class documents the items in the returned tuples.

andy-slac committed:
Allow Query.general to take positional parameters with field names.

andy-slac committed:
Add unit test for general queries and fix serialization of ingest_date.

Adding unit tests exposed the issue with serialization of ingest_time.
We still create schema with a native time type for that column, while
its corresponding ColumnSpec says it should be astropy time. The code
that I added recently used SerializableTime with type adapter to
serialize it which did not work with datetime. I had to reimplement
serialization methods to allow more flexible handling of ingest_date types.

andy-slac committed:
Simplify construction of DatasetAssociation from general query result.

andy-slac commented on python/lsst/daf/butler/remote_butler/_registry.py:
```diff
@@ -513,7 +513,19 @@ def queryDatasetAssociations(
+        # queryCollections only accepts DatasetType.
+        if isinstance(datasetType, str):
+            datasetType = self.getDatasetType(datasetType)
+        resolved_collections = self.queryCollections(
+            collections, datasetType=datasetType, collectionTypes=collectionTypes, flattenChains=flattenChains
+        )
+        with self._butler._query() as query:
+            query = query.join_dataset_search(datasetType, resolved_collections)
+            result = query.general(
+                datasetType.dimensions,
+                dataset_fields={datasetType.name: {"dataset_id", "run", "collection", "timespan"}},
+            )
+            yield from DatasetAssociation.from_query_result(result, datasetType)
```
@timj, this is how replacement code for `queryDatasetAssociations` is going to look like. I checked `RepoExportContext` class to see if I can replace `queryDatasetAssociations` usage there, but that class does not even have butler instance now, and it uses registry for everything. We can do it later and keep `queryDatasetAssociations` in registry for now. Both direct and remote registries use new query system.

timj replied:
Great. I had forgotten that we need to change the export code so that it doesn't use registry.

TallJimbo commented on python/lsst/daf/butler/queries/_general_query_results.py:
```diff
@@ -83,33 +102,36 @@ def __iter__(self) -> Iterator[dict[str, Any]]:
-            ref = DatasetRef(dataset_type, data_id, row[run_key], id=row[id_key])
-            yield ref, row
+            values = tuple(row[key] for key in all_dimensions.required)
+            data_coordinate = DataCoordinate.from_required_values(all_dimensions, values)
```
`GeneralResultSpec` guarantees that the full dimensions are available.  So we can and probably should extract the full data IDs here (and below for the dataset data IDs).

TallJimbo commented on python/lsst/daf/butler/queries/_query.py:
```diff
@@ -329,10 +346,32 @@ def general(
+                    # Could be because someone asked for the key field.
+                    # Potentially expand dimensions (but tree needs to have
+                    # them already or there will be an error).
+                    dimensions = dimensions.union(dimension.minimal_group)
```
Instead of updating a `DimensionGroup` variable as you go, I'd recommend accumulating a `set` of dimension names and doing `universe.conform(names | dimension_fields_dict.keys())` at the end.  And then you can do something like https://github.com/lsst/daf_butler/blob/8e576370746fa86e5d065499e4f114fb97b83289/python/lsst/daf/butler/queries/_query.py#L200-L201 to join in the missing dimensions.  That will make this a lot easier to use, as well as consistent with how `where` handles dimensions that aren't in the query yet.

TallJimbo commented on python/lsst/daf/butler/column_spec.py:
```diff
@@ -88,6 +89,102 @@
+        if value is None:
+            return None
+        elif isinstance(value, str):
+            return datetime.datetime.fromisoformat(value)
```
I understand wanting to accept `datetime.datetime` and `str` from users and databases for backwards compatibility, but I don't understand why we'd want to return a `datetime.datetime` to users or serialize to `str` going forward.

My recollection is that we wanted to eventually switch to astropy/nsec for ingest_time everywhere, but it's a tricky migration due to the timezone stuff and we don't really have a sound plan for doing it.  But I'd like to get the in-memory Python types consistent.

If that's all true, I think we could definitely support coercion from `str` or `datetime.datetime` in `pydantic_utils._deserialize_time`, and that might be enough to replace the new serialization dispatch logic here.  But it seems likely I'm not thinking about at least one of the places where we need to do the time type conversions.

andy-slac replied:
I chose this for compatibility between direct butler and remote butler. Queries with direct butler return whatever sqlalchemy does for that column, which is `datetime` in current schema. It may be surprising for clients if remote butler returned astropy time instead for the same query. We could convert all `datetime` instances returned from queries to astropy, then serialization would need to handle only one type of time. 
These new methods are only used to (de)serialize query results, I'm not sure what happens when the query itself includes time instances in `bind`, need to check.

TallJimbo replied:
Thus far it hasn't actually been possible to get the `ingest_date` back from a query - there's no interface in the Registry query system that can do it, and the "general" result type is the only way to do it in the new one.  So I think we've still got until we open up `butler._query*` as a public interface to get this this fixed, and IMO that means coercing to astropy.  I'm happy for that to be another ticket, though.

andy-slac replied:
OK, I'll make a new ticket for that.

andy-slac committed:
Update RepoExportContext to use new query system instead of queryDatasetAssociations.
